common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -112,34 +116,48 @@ public ShuffleIndexInformation load(File file) throws IOException {
     this.errorHandler = new ErrorHandler.BlockPushErrorHandler();
   }
 
+  private AppShuffleInfo validateAndGetAppShuffleInfo(String appId) {
+    // TODO: [SPARK-33236] Change the message when this service is able to handle NM restart
+    AppShuffleInfo appShuffleInfo =
+      Preconditions.checkNotNull(appsShuffleInfo.get(appId),
+        "application " + appId + " is not registered or NM was restarted.");
+    return appShuffleInfo;
+  }
+
   /**
    * Given the appShuffleId and reduceId that uniquely identifies a given shuffle partition of an
    * application, retrieves the associated metadata. If not present and the corresponding merged
    * shuffle does not exist, initializes the metadata.
    */
   private AppShufflePartitionInfo getOrCreateAppShufflePartitionInfo(
-      AppShuffleId appShuffleId,
+      AppShuffleInfo appShuffleInfo,
+      int shuffleId,
       int reduceId) {
-    File dataFile = getMergedShuffleDataFile(appShuffleId, reduceId);
-    if (!partitions.containsKey(appShuffleId) && dataFile.exists()) {
+    File dataFile = appShuffleInfo.getMergedShuffleDataFile(shuffleId, reduceId);
+    ConcurrentMap<Integer, Map<Integer, AppShufflePartitionInfo>> partitions =
+      appShuffleInfo.partitions;
+    if (!partitions.containsKey(shuffleId) && dataFile.exists()) {
       // If this partition is already finalized then the partitions map will not contain
       // the appShuffleId but the data file would exist. In that case the block is considered late.
       return null;
     }
```

Comment:
nit: Do we want to combine the validation in a thread safe way ?

```
    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
            partitions.compute(shuffleId, (id, map) -> {
              if (null == map) {
                // If this partition is already finalized then the partitions map will not contain
                // the appShuffleId but the data file would exist. In that case the block is considered late.
                if (dataFile.exists()) {
                  return null;
                }
                return Maps.newConcurrentMap();
              } else {
                return map;
              }
            });

    if (null == shufflePartitions) {
      return null;
    }
```

In my opinion, I dont think this is buying us much ... but wanted to suggest anyway, since it is a stricter check.

Reply:
Updated.